// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"sastoj/ent/predicate"
	"sastoj/ent/problemcase"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// ProblemCaseDelete is the builder for deleting a ProblemCase entity.
type ProblemCaseDelete struct {
	config
	hooks    []Hook
	mutation *ProblemCaseMutation
}

// Where appends a list predicates to the ProblemCaseDelete builder.
func (pcd *ProblemCaseDelete) Where(ps ...predicate.ProblemCase) *ProblemCaseDelete {
	pcd.mutation.Where(ps...)
	return pcd
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (pcd *ProblemCaseDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, pcd.sqlExec, pcd.mutation, pcd.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (pcd *ProblemCaseDelete) ExecX(ctx context.Context) int {
	n, err := pcd.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (pcd *ProblemCaseDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(problemcase.Table, sqlgraph.NewFieldSpec(problemcase.FieldID, field.TypeInt64))
	if ps := pcd.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, pcd.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	pcd.mutation.done = true
	return affected, err
}

// ProblemCaseDeleteOne is the builder for deleting a single ProblemCase entity.
type ProblemCaseDeleteOne struct {
	pcd *ProblemCaseDelete
}

// Where appends a list predicates to the ProblemCaseDelete builder.
func (pcdo *ProblemCaseDeleteOne) Where(ps ...predicate.ProblemCase) *ProblemCaseDeleteOne {
	pcdo.pcd.mutation.Where(ps...)
	return pcdo
}

// Exec executes the deletion query.
func (pcdo *ProblemCaseDeleteOne) Exec(ctx context.Context) error {
	n, err := pcdo.pcd.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{problemcase.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (pcdo *ProblemCaseDeleteOne) ExecX(ctx context.Context) {
	if err := pcdo.Exec(ctx); err != nil {
		panic(err)
	}
}
